mirgecom/artificial_viscosity.py

```diff
@@ -0,0 +1,196 @@
+r""":mod:`mirgecom.artificial viscosity` Artificial viscocity for Euler.
+
+Artificial viscosity for the Euler Equations:
+
+.. math::
+
+    \partial_t \mathbf{R} = \nabla\cdot{\varepsilon\nabla\mathbf{R}}
+
+where:
+
+-  state $\mathbf{R} = [\rho, \rho{E}, \rho\vec{V}, \rho{Y}_\alpha]$
+-  artifical viscosity coefficient $\varepsilon$
+
+To evalutate the second order derivative the problem is recast as a set of first
+ order problems:
+
+.. math::
+
+    \partial_t \mathbf{R} = \nabla\cdot{\mathbf{Q}}
+    \mathbf{Q} = \varepsilon\nabla\mathbf{R}
+
+where $\mathbf{Q}$ is an intermediate variable.
+
+RHS Evaluation
+^^^^^^^^^^^^^^
+
+.. autofunction:: artificial_viscosity
+"""
+
+__copyright__ = """
+Copyright (C) 2020 University of Illinois Board of Trustees
+"""
+
+__license__ = """
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+import numpy as np
+from pytools.obj_array import (
+    obj_array_vectorize,
+    obj_array_vectorize_n_args,
+)
+from meshmode.dof_array import thaw
+from meshmode.mesh import BTAG_ALL, BTAG_NONE  # noqa
+from grudge.eager import interior_trace_pair, cross_rank_trace_pairs
+from grudge.symbolic.primitives import TracePair
+from mirgecom.tag_cells import smoothness_indicator
+
+
+def _facial_flux_r(discr, q_tpair):
+
+    actx = q_tpair.int.array_context
+
+    flux_dis = q_tpair.avg
+
+    normal = thaw(actx, discr.normal(q_tpair.dd))
+
+    flux_out = flux_dis * normal
+
+    return discr.project(q_tpair.dd, "all_faces", flux_out)
```

Comment:
Even though this is just an internal function, I am a huge stickler for at least a bit of exposition (documentation), especially for people returning to this code in the future. I think adding a short docstring stating that you are discretizing using central difference numerical fluxes (averaging) would be nice to have here. Similarly below for the numerical flux for the auxiliary equation.

Reply:
How's this grab you guys? w-hagen/mirgecom#13